### DISCUSSION TITLE
`bitcoin-internals` got snavelled

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#1702] [Announcements]
Some f***er snavelled `bitcoin-internals` on crates.io using our repo, what's up with that? 

https://crates.io/users/marspoolxyz

**Comment by dpc** [comment]
Account looks suspicious. Issues disabled so can't ask, mostly forks and empty repos. Looks malicious. Reported on github, but no smoking gun here yet.

I downloaded the crate with `cargo crev goto -u bitcoin-internals` and don't see anything immediately malicious. I guess a `build.rs` would be most natural place to put something, but looks normal.

**Reply by apoelstra** [reply]
Definitely looks malicious. Multiple repos, all of which are typo-squatting.

**Reply by dpc** [reply]
Do you want to ask what's going on? https://github.com/marspoolxyz/dappMarsPool/issues

Other than that reporting to github and crates.io seems like the only actionable item, but without some evidence of maliciousness, I'm not sure how much can be done.

**Reply by apoelstra** [reply]
Uploading a crate with metadata pointing to our org, without our permission or knowledge, is evidence of maliciousness. I have sent an email to crates.io with this information. We can certainly open an issue on their github page as well, assuming this github person is related to the crates.io person.

**Reply by apoelstra** [reply]
If you, or anyone, is on Zulip or Discord, crates.io also has help channels there. I think our best bet is to focus on the fraudulent metadata, since crates.io explicitly has a "no such thing as squatting, first come first serve" policy, and they will simply blame us for not claiming the crate early.

Obviously, I'd like to have the name back, but if we can convince them to just yank the crate and perma-ban the name I'd be okay with that too.

**Comment by apoelstra** [comment]
Wow, we would've actually gotten (briefly) nailed by this had we done the release with a valid version number for our `bitcoin-internals`, since crates would've silently used the squatter's copy. Had we then forgotten that we needed to publish ours, we might not have noticed for a long time.

**Comment by apoelstra** [comment]
I shot an email to help@crates.io

```
Helo,


The crate bitcoin-internals https://crates.io/crates/bitcoin-internals was uploaded 14 days ago by the user "Marspoolxyz". The repository and documentation link to those belonging to the rust-bitcoin org. Marspoolxyz is not, and never has been, a member of this org, and the crate precludes one of our own by that name (an upcoming release of rust-bitcoin depends on a local copy, and very nearly depends on the remote one).

I appreciate that crates.io does not define legitimacy or have squatting rules. But this crate has metadata which is clearly fraudulent.

The existing uploaded data is identical to that at https://github.com/rust-bitcoin/rust-bitcoin/tree/master/internals commit 10eb0da1eff1bb120dd31db40d8695716268879d except that this person has added himself as an author. We use a permissive CC0 license which technically allows this person to do this, but it is still fraudulent and I hope that crates.io does the correct thing by removing this crate and banning this person.

Further evidence of malice is this person's other lists of crates, which include "litcoin" (typosquatting on litecoin) as well as several other litecoin-themed crates which have similarly deceptive metadata.

rust-bitcoin is a security and cryptography-focused crate and we view the presence of this crate as a security risk.


Best
Andrew
```

**Comment by apoelstra** [comment]
Assuming we don't win any crates.io dispute, I guess we need to come up with a new name (and publish it fast :)). It should have a fair bit of distance from "internals" so that the fraudulent crate isn't typo-squatting it. So, like, I don't think we should drop the `s` or change `-` to `_` (or both :P).

**Reply by Kixunil** [reply]
Let's discuss this over e-mail so there's no more attack.

**Comment by apoelstra** [comment]
I realize actually that the author directly contacted me a couple weeks ago with some questions about litecoin. I think he may be legitimately confused and have done this by accident (fingers crossed). I will contact him.

**Reply by apoelstra** [reply]
```
Subject: Re: Hello
Reply-To: 
In-Reply-To: <CAPnne128Cr5vJDzzXO4XoZ=rduRVYjaPpibOcNoxySAq-mE3vQ@mail.gmail.com>


Hey Waheed,


It looks like you uploaded bitcoin-internals under that name -- I guess that
you meant to upload it as litecoin-internals? We actually need the name
bitcoin-internals for our own project. :)

It is this crate: https://crates.io/crates/bitcoin-internals

Would you be willing to transfer ownership of that name to me (apoelstra)
so that we can upload our copy?


Best
Andrew
```

**Comment by Kixunil** [comment]
Shit, I was thinking not so long ago that I should've grabbed it just in case but didn't actually do it. I will remember this the next time.

Sadly if crates.io won't resolve these cases it'll put pressure on people to grab many interesting names. :(

**Reply by dpc** [reply]
We need namespaces. It's ridiculous for a bigger project with multiple crates that anyone can just start getting crates prefixed with the same name.

Each user should get `@username` namespace and all crates would go into `@username/crate`, and non-@-prefixed namespaces should be given away on request after (brief, but still) human review.

Or something like this.

**Reply by Kixunil** [reply]
Then people will typo-squat usernames instead. That being said I quite like the recent idea that if you publish `bitcoin` crate you get `bitcoin/*` namespace.

**Reply by dpc** [reply]
> Then people will typo-squat usernames instead. 

And that's OK. 100% typo-squatting protection would require either (lots of) human review of everything or AI, etc. Important projects would get promoted to where typo-squatting is no longer possible, striking a good balance between "anyone can publish" and "typo-squatting protection" with minimum human intervention required.

**Comment by Kixunil** [comment]
FYI, this is still problematic. I just wrote `bitcoin-internals` out of habit. Thankfully I realized before even compiling the code. Ideally there would be some cargo option to perform hard ban.

**Comment by tarcieri** [comment]
FYI, there is not yet an automated process for the crates.io team to transfer a crate from one user to another, even if they decided they wanted to: https://github.com/rust-lang/rfcs/pull/2614

Your best bets are either to politely request a crate transfer, or if that doesn't work, to pick a new name

**Reply by tcharding** [reply]
No worries, thanks @tarcieri

**Reply by Kixunil** [reply]
Yeah, that's what we did. The presence of the crate still worries me a bit since I got used to the name.

**Comment by apoelstra** [comment]
Marspoolxyz got back to me and transferred `bitcoin-internals` into my control :).

I guess we don't want to use it now, but we can update its description on crates.io at least to point people to bitcoin-private.

**Reply by Kixunil** [reply]
Oh, that's great! Given we use the word `internals` everywhere in the codebase how about do it the other way around and keep `internals` instead?

**Reply by apoelstra** [reply]
Sure, sounds good to me!

**Reply by bjorn3** [reply]
Last time I checked crates.io doesn't show pending ownership invites anywhere.

Edit: Still the case. See https://github.com/rust-lang/crates.io/issues/2868

**Reply by tcharding** [reply]
Ah sweet, props to Marspoolxyz.
